ocaml/xapi/xapi_pool_transition.ml

```diff
@@ -131,15 +131,23 @@ let attempt_two_phase_commit_of_new_master ~__context (manual: bool) (peer_addre
     error "Some hosts failed to commit [ %s ]: this node will now restart in a broken state"
       (String.concat "; " !hosts_which_failed);
     (* Immediately just in case *)
-    Db_cache_impl.flush_and_exit (Db_connections.preferred_write_db ()) Xapi_globs.restart_return_code
+    let dbconn = Db_connections.preferred_write_db in
+    Xapi_stdext_monadic.Opt.iter (fun db ->
+      Db_cache_impl.flush_and_exit db Xapi_globs.restart_return_code
+    ) dbconn
   end;
   (* If this is an automatic transition then there is no other master to clash with and so
      we can restart immediately. NB if we are the master (and this code is being used to assert
      our authoritah on some slaves then we shouldn't restart otherwise we'll get into a restart loop. *)
   if not(manual) then
     if am_master_already
     then info "Not restarting since we are the master already"
-    else Db_cache_impl.flush_and_exit (Db_connections.preferred_write_db ()) Xapi_globs.restart_return_code;
+    else begin
+      let dbconn = Db_connections.preferred_write_db in
+      Xapi_stdext_monadic.Opt.iter (fun db ->
```

Comment:
I think we should rather fail here if we have no DB connections, `flush_and_exit` is meant to exit XAPI, and if we have no DB connections we would silently not exit XAPI.